### PR TITLE
lora/lorawan: better errors and provisioning

### PR DESCRIPTION
**Pull request by deadprogram**
This PR extracts all of the errors into a single place, also removing unneeded heap allocations.

The LoRaWAN `atcmd` examples should now output to the console after `AT+JOIN` all of the information needed to simply send from that device in the future without joining again, as long as the same keys are used.

### CODE REVIEW

File: lora/lorawan/otaa.go [hunk 1]
```diff
@@ -4,16 +4,15 @@ import (
...
 )
 
 // Otaa is used to store Over The Air Activation data of a LoRaWAN session
 type Otaa struct {
 	DevEUI   [8]uint8
 	AppEUI   [8]uint8
 	AppKey   [16]uint8
-	DevNonce [2]uint8
-	AppNonce [3]uint8
+	devNonce [2]uint8
```
Comment (ofauchon) [inline lora/lorawan/otaa.go:14]: You made these two variable private 
Should we do the same in a future PR, and implement Getters when needed ?

**Review by ofauchon — COMMENTED** [review]
Looks good to me !

**Comment by ofauchon** [comment]
Ok works on LoraE5...


```
+ID: AppEui, 0000000000000000
+ID: DevEui, a84041000181b365
+KEY: APPKEY, <hiddensecret>
+JOIN: Starting
+JOIN: no JoinAccept packet received
+ID: AppEui, 0000000000000000
+ID: DevEui, a84041000181b365
+KEY: APPKEY, <hiddensecret>
+JOIN: Starting
+JOIN: Network joined
+JOIN: DevEui, a84041000181b365
+JOIN: AppEui, 0000000000000000
+JOIN: DevAddr, <hiddensecret>
+JOIN: NetID, 000000
+JOIN: NwkSKey, <hiddensecret>
+JOIN: AppSKey, <hiddensecret>
+JOIN: Done



....
Errors handling as well....

+ID: invalid EUI length
+ID: invalid EUI length
+ID: invalid EUI length
+ID: invalid EUI length
+KEY: invalid AppKey length

```

File: lora/lorawan/otaa.go (hunk 1 above)
Comment (deadprogram) [reply lora/lorawan/otaa.go:14]: Good idea!

**Comment by deadprogram** [comment]
I have made some pretty large refactoring changes, but now I am done.

The interface to the sx126x should be a lot clearer now, and also not having to re-implement any interrupt code should make it a lot easier for users.

@ofauchon please take a look!

File: examples/lora/lorawan/common/sx126x.go [hunk 2]
```diff
@@ -23,23 +19,28 @@ var (
...
+)
+
+func newRadioControl() sx126x.RadioController {
+	return sx126x.NewRadioControl(nssPin, busyPin, dio0Pin, dio1Pin, rxPin, txLowPin, txHighPin)
+}
+
 // do sx126x setup here
 func SetupLora() (lora.Radio, error) {
-	loraRadio = sx126x.New(machine.SPI3)
+	loraRadio = sx126x.New(spi)
```
Comment (ofauchon) [inline examples/lora/lorawan/common/sx126x.go:34]: Why not moving SPI configuration in radio controller as well ?

Comment (deadprogram) [reply examples/lora/lorawan/common/sx126x.go:34]: I just found that part a bit less obvious, also hoping to get some mock-based testing in place in a future iteration.

**Review by ofauchon — APPROVED** [review]
Code looks good to me. 
I could test this version with atcmd and basic-demo example on Lora E5. 
Join and uplink still working.

**Comment by ofauchon** [comment]
Two latest commits tested again on LoraE5 dev board. 
atcmd and basic-demo example still working (join and uplink)

LGTM

**Comment by deadprogram** [comment]
Merging, and then we can make further changes from there. Thank you very much for all your help with this @ofauchon !!!

**Pull request merged**